File: src/leetcode/L2.java
package leetcode;


/**
 * 给出两个 非空 的链表用来表示两个非负的整数。其中，它们各自的位数是按照 逆序 的方式存储的，
 * 并且它们的每个节点只能存储 一位 数字。
 *
 * 如果，我们将这两个数相加起来，则会返回一个新的链表来表示它们的和。
 *
 * 您可以假设除了数字 0 之外，这两个数都不会以 0 开头。
 */
public class L2 {

    public static void main(String[] args) {
        ListNode listNode1 = new ListNode(1);
        listNode1.next = new ListNode(9);

        ListNode listNode2 = new ListNode(1);
        listNode2.next = new ListNode(9);

        L2 l2 = new L2();
        ListNode listNode = l2.addTwoNumbers(listNode1, listNode2);

        System.out.println("最终结果：");
        while(listNode != null)
        {
            System.out.print(listNode.val);
            listNode = listNode.next;
        }
    }

    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummyHead = new ListNode(0);
        ListNode p = l1, q = l2, curr = dummyHead;
        int carry = 0;
        while (p != null || q != null) {
            int x = (p != null) ? p.val : 0;
            int y = (q != null) ? q.val : 0;
            int sum = carry + x + y;
            System.out.println("sum:" + sum + " carry:" + carry + " x:" + x + " y:" + y);
            carry = sum / 10;
            System.out.println("carry: " + carry);
            curr.next = new ListNode(sum % 10);

            curr = curr.next;
            if (p != null) p = p.next;
            if (q != null) q = q.next;
        }
        if (carry > 0) {
            curr.next = new ListNode(carry);
        }
        return dummyHead.next;
    }
}

class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}

File: src/leetcode_kotlin/L2.kt
package leetcode_kotlin

class L2 {

    fun addTwoNumbers(l1:ListNode,l2:ListNode):ListNode? {
        var dummyHead = ListNode(0)

        var p:ListNode? = l1
        var q:ListNode?= l2
        var curr = dummyHead

        var carry = 0
        while(p!=null || q != null){
            var x = 0
            if(p != null) {
                x = p.value
            }
            var y = 0
            if(q != null) {
                y = q.value
            }
            var sum = carry + x + y

            println("sum:$sum carry:$carry x:$x y:$y")
            carry = sum / 10
            print("carry:$carry")
            curr.next = ListNode(sum%10)

            curr = curr.next!!
            if(p != null) {
                p = p.next
            }
            if(q!= null) {
                q = q.next
            }

            println("while ")
        }
        if (carry > 0) {
            curr.next = ListNode(carry)
        }
        return dummyHead.next
    }

}


fun main(args: Array<String>) {
    var listNode1 = ListNode(1)
    listNode1.next = ListNode(9)

    var listNode2 = ListNode(1)
    listNode2.next = ListNode(9)

    val l2 = L2()
    var listNode = l2.addTwoNumbers(listNode1, listNode2)
    println("结果：")
    while (listNode != null) {
        print(listNode.value)
        listNode = listNode.next
    }



}

class ListNode constructor(x: Int) {

    var value: Int = 0
    var next: ListNode? = null

    init {
        this.value = x
    }

}


